prov/rxd/src/rxd_ep.c

```diff
@@ -0,0 +1,1789 @@
+	}
+out:
+	rxd_ep_unlock_if_required(rxd_ep);
+	return ret;
+}
+
+static ssize_t rxd_ep_recv(struct fid_ep *ep, void *buf, size_t len, void *desc,
+			    fi_addr_t src_addr, void *context)
+{
+	struct fi_msg msg;
+	struct iovec msg_iov;
+	memset(&msg, 0, sizeof(msg));
+	msg_iov.iov_base = buf;
+	msg_iov.iov_len = len;
+
+	msg.msg_iov = &msg_iov;
+	msg.desc = &desc;
+	msg.iov_count = 1;
+	msg.addr = src_addr;
+	msg.context = context;
+	msg.data = 0;
+	return rxd_ep_recvmsg(ep, &msg, RXD_USE_OP_FLAGS);
+}
+
+static ssize_t rxd_ep_recvv(struct fid_ep *ep, const struct iovec *iov, void **desc,
+			     size_t count, fi_addr_t src_addr, void *context)
+{
+	struct fi_msg msg;
+	memset(&msg, 0, sizeof(msg));
+	msg.msg_iov = iov;
+	msg.desc = desc;
+	msg.iov_count = count;
+	msg.addr = src_addr;
+	msg.context = context;
+	msg.data = 0;
+	return rxd_ep_recvmsg(ep, &msg, RXD_USE_OP_FLAGS);
+}
+
+static inline void *rxd_mr_desc(struct fid_mr *mr, struct rxd_ep *ep)
+{
+	return (ep->do_local_mr) ? fi_mr_desc(mr) : NULL;
+}
+
+int rxd_ep_repost_buff(struct rxd_rx_buf *buf)
+{
+	int ret;
+	ret = fi_recv(buf->ep->dg_ep, buf->buf, buf->ep->domain->max_mtu_sz,
+		      rxd_mr_desc(buf->mr, buf->ep),
+		      FI_ADDR_UNSPEC, &buf->context);
+	if (ret)
+		FI_WARN(&rxd_prov, FI_LOG_EP_CTRL, "failed to repost\n");
+	return ret;
+}
+
+static int rxd_ep_set_conn_id(struct rxd_ep *ep)
+{
+	int ret;
+	ep->name = calloc(1, ep->addrlen);
+	if (!ep->name)
+		return -FI_ENOMEM;
+
+	ret = fi_getname(&ep->dg_ep->fid, ep->name, &ep->addrlen);
+	if (ret)
+		return -FI_EINVAL;
+
+	ret = rxd_av_dg_reverse_lookup(ep->av, 0, ep->name, ep->addrlen,
+					&ep->conn_data);
+	if (!ret)
+		ep->conn_data_set = 1;
+	return 0;
+}
+
+int rxd_ep_enable(struct rxd_ep *ep)
+{
+	ssize_t i, ret;
+	struct fid_mr *mr = NULL;
+	struct rxd_rx_buf *rx_buf;
+
+	ret = fi_enable(ep->dg_ep);
+	if (ret)
+		return ret;
+
+	rxd_ep_lock_if_required(ep);
+	ret = rxd_ep_set_conn_id(ep);
+	if (ret)
+		goto out;
+
+	ep->credits = ep->rx_size;
+	for (i = 0; i < ep->rx_size; i++) {
+		rx_buf = ep->do_local_mr ?
+			util_buf_get_ex(ep->rx_pkt_pool, (void **)&mr) :
+			util_buf_get(ep->rx_pkt_pool);
+
+		if (!rx_buf) {
+			ret = -FI_ENOMEM;
+			goto out;
+		}
+
+		rx_buf->mr = mr;
+		rx_buf->ep = ep;
+		ret = rxd_ep_repost_buff(rx_buf);
+		if (ret)
+			goto out;
+		slist_insert_tail(&rx_buf->entry, &ep->rx_pkt_list);
+	}
+
+	fastlock_acquire(&ep->domain->lock);
+	dlist_insert_tail(&ep->dom_entry, &ep->domain->ep_list);
+	fastlock_release(&ep->domain->lock);
+	ret = 0;
+out:
+	rxd_ep_unlock_if_required(ep);
+	return ret;
+}
+
+struct rxd_peer *rxd_ep_getpeer_info(struct rxd_ep *ep, fi_addr_t addr)
+{
+	return &ep->peer_info[addr];
+}
+
+void rxd_ep_lock_if_required(struct rxd_ep *ep)
+{
+	/* todo: do locking based on threading model */
+	fastlock_acquire(&ep->lock);
+}
+
+void rxd_ep_unlock_if_required(struct rxd_ep *ep)
+{
+	/* todo: do unlocking based on threading model */
+	fastlock_release(&ep->lock);
+}
```

Comment:
The above functions looks like they could be in common code. Is there a plan to move them there once the todo is done?